rascal-lsp/src/main/rascal/lang/rascal/lsp/IDECheckerWrapper.rsc

```diff
@@ -0,0 +1,96 @@
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+}
+module lang::rascal::lsp::IDECheckerWrapper
+
+import IO;
+import List;
+import Set;
+import String;
+import analysis::graphs::Graph;
+import util::FileSystem;
+import util::Reflective;
+
+import lang::rascal::\syntax::Rascal;
+import lang::rascalcore::check::Checker;
+import lang::rascalcore::check::ModuleLocations;
+
+@synopsis{
+    This function is a wrapper for the Rascal type checker, taking into account that dependencies
+    may be open in the current workspace. It takes an overapproximation of the import/extend graph,
+    and calls the type checker on (transitive) dependencies first.
+
+    This function must only be used in an IDE context.
+}
+list[ModuleMessages] checkFile(loc l, node(loc file) getParseTree, PathConfig(loc file) getPathConfig) {
+    checkForImports = [pt | start[Module] pt := getParseTree(l)];
+    checkedForImports = {};
+
+    rel[loc, loc] dependencies = {<root, root> | root := inferProjectRoot(l)};
+    
+    msgs = [];
+
+    while (tree <- checkForImports) {
+        currentSrc = tree.src.top;
+        currentProject = inferProjectRoot(currentSrc);
+        for (i <- tree.top.header.imports) {
+            ml = getRascalModuleLocation("<i.\module>", getPathConfig(currentProject));
+            mlProject = inferProjectRoot(ml);
+            dependencies += <currentProject, resolveLocation(mlProject)>;
+            if (ml.extension == "rsc" && ml notin checkedForImports, start[Module] mlpt := getParseTree(ml)) {
+                checkForImports += mlpt;
+            }
+        }
+        checkedForImports += currentSrc;
+        checkForImports -= tree;
+    }
+    modulesPerProject = classify(checkedForImports, loc(loc l) {return inferProjectRoot(l);});
+    return [*check([*modulesPerProject[project]], makeCompilerConfig(getPathConfig(project))) | project <- reverse(order(dependencies)), project in modulesPerProject];
+}
+
+RascalCompilerConfig makeCompilerConfig(PathConfig pcfg) = rascalCompilerConfig(pcfg[resources=pcfg.bin]);
+
+loc inferProjectRoot(loc member) {
+    current = member;
+    if (!isDirectory(current)) {
+        current = current.parent;
+    }
+
+    while (exists(current), isDirectory(current)) {
+        if (exists(current + "META-INF" + "RASCAL.MF")) {
+            return ensureTrailingSpaceInPath(current);
+        }
+        if (!current.parent?) {
+            return isDirectory(member) ? member : member.parent;
+        }
+        current = current.parent;
```

Comment:
what if `current.parent == current`?

Reply:
Doesn't Rascal always throw a `NoParent` instead of returning the location itself?

Reply:
The condition in the `if` 'catches' the `NoParent`s

Reply:
aaaah, right, my bad.